src/framework/audio/internal/sequencer.cpp

```diff
@@ -0,0 +1,291 @@
+
+    std::function<void(const track_t&)> applyStaus = [](const track_t&) {};
+    switch (m_status) {
+    case PLAYING:
+        applyStaus = [](const track_t& track) { track->play(); };
+        m_clock->start();
+        break;
+    case STOPED:
+        applyStaus = [](const track_t& track) { track->stop(); };
+        m_clock->stop();
+        break;
+    }
+
+    for (auto& track : m_tracks) {
+        applyStaus(track.second);
+    }
+}
+
+std::optional<Sequencer::track_t> Sequencer::track(Sequencer::track_id id) const
+{
+    if (m_tracks.find(id) != m_tracks.end()) {
+        return m_tracks.at(id);
+    }
+    return std::nullopt;
+}
+
+Sequencer::midi_track_t Sequencer::midiTrack(Sequencer::track_id id) const
+{
+    if (auto track_var = track(id)) {
+        if (auto track_value = track_var.value_or(nullptr)) {
+            return std::dynamic_pointer_cast<IMIDIPlayer>(track_value);
+        }
+    }
+    return nullptr;
+}
+
+Sequencer::midi_track_t Sequencer::createMIDITrack(ISequencer::track_id id)
+{
+    auto player = std::make_shared<MIDIPlayer>();
+    m_tracks[id] = player;
+    return player;
+}
+
+Sequencer::audio_track_t Sequencer::createAudioTrack(Sequencer::track_id id)
+{
+    auto player = std::make_shared<AudioPlayer>();
+    m_tracks[id] = player;
+    m_audioTrackAdded.send(player);
+    return player;
+}
+
+Sequencer::audio_track_t Sequencer::audioTrack(Sequencer::track_id id) const
+{
+    if (auto track_var = track(id)) {
+        if (auto track_value = track_var.value_or(nullptr)) {
+            return std::dynamic_pointer_cast<IAudioPlayer>(track_value);
+        }
+    }
+    return nullptr;
+}
+
+std::shared_ptr<IMIDIPlayer> Sequencer::instantlyPlayMidi(const midi::MidiData& data)
+{
+    if (!data.isValid()) {
+        return nullptr;
+    }
+    auto player = std::make_shared<MIDIPlayer>();
+    auto midiStream = std::make_shared<midi::MidiStream>();
+    std::lock_guard guard(m_syncMutex);
+
+    midiStream->initData = data;
+    midiStream->isStreamingAllowed = false;
+    midiStream->lastTick = data.lastChunksTick();
+    player->loadMIDI(midiStream);
+    player->play();
+
+    m_backgroudPlayers.push_back({ 0 /*ms*/, player });
+    return player;
+}
+
+void Sequencer::buildRpcReflection()
+{
+    using namespace rpc;
+    m_rpcReflection = {
+        { "play",    [this](ArgumentList) -> Variable { return play(); } },
+        { "pause",   [this](ArgumentList) -> Variable { pause(); return {}; } },
+        { "stop",    [this](ArgumentList) -> Variable { stop(); return {}; } },
+        { "seek",    [this](ArgumentList args) -> Variable { seek(args[0].toUInt()); return {}; } },
+        { "rewind",  [this](ArgumentList)      -> Variable { rewind(); return {}; } },
+        { "setLoop", [this](ArgumentList args) -> Variable { setLoop(args[0].toUInt(), args[1].toUInt()); return {}; } },
+        { "unsetLoop",  [this](ArgumentList)   -> Variable { unsetLoop(); return {}; } }
+    };
```

Comment:
2. The second problem is the implementation of subscription to channels for a given argument, such as
```
async::Channel<mu::midi::tick_t> midiTickPlayed(track_id id) const
```

The target class, Sequencer, just has a method that returns a channel, but the returned channel cannot be returned as a result of executing the method, it won't work.
Also, if the RPC system requires serialization of data, then the channel as it is also cannot be simply transferred through such a system. Serialization or rather emulation of the channel is required.

Reply:
fixed